### DISCUSSION TITLE
Allow configuration to DISABLE updating the updated_at timestamp during mass update?

### DISCUSSION THREAD

**Discussion by stevenmaguire** [iksaku/laravel-mass-update#5] [Ideas]
A common use case of mine is to perform mass updates to models where you DO NOT want to touch the updated_at timestamp. Are you open to supporting that use case with this library?

**Comment by iksaku** [comment]
Hello Steven. I intend for this method to stick as close as possible to Laravel's `Model::query()->update()` method. As natively there's no option to tell the query builder not to touch `UPDATED_AT` timestamp, I think I'll refrain from providing that option myself.

**Reply by stevenmaguire** [reply]
I appreciate where you're coming from with that. I think it is worth noting that Laravel's default behavior allows (WITH CONFIGURATION) the very behavior I am asking about. For instance...

https://laravel.com/docs/9.x/eloquent#muting-events

<img width="745" alt="CleanShot 2022-08-08 at 18 00 56@2x" src="https://user-images.githubusercontent.com/1851973/183528841-c015de00-bf48-43b0-8675-ebbda48f9ff9.png">


https://laravel.com/docs/9.x/eloquent#saving-a-single-model-without-events

<img width="749" alt="CleanShot 2022-08-08 at 18 00 49@2x" src="https://user-images.githubusercontent.com/1851973/183528803-9e689937-d227-4123-bfe0-4777427a3393.png">

https://github.com/laravel/framework/blob/213074f15e6398137e0b7cf1809368e62b0c03f1/src/Illuminate/Database/Eloquent/Model.php#L350

The model itself is capable of being told to ignore touch during save.

<img width="716" alt="CleanShot 2022-08-08 at 18 07 49@2x" src="https://user-images.githubusercontent.com/1851973/183529556-409913b3-e8a0-4b7e-804a-503c978599f7.png">

----------------------------------

I am not advocating for the DEFAULT behavior to change. I am advocating for the behavior to be available with configuration.

**Reply by iksaku** [reply]
Doing configuration-based behavior is a good idea, not going to lie, however, ignoring timestamp updates dynamically is still not a feature with Laravel.

The code provided in `Model::isIgnoringTouch()` is intended for use with relationships, and if we rely on such function, unexpected behavior will arise sooner or later:
https://github.com/laravel/framework/blob/213074f15e6398137e0b7cf1809368e62b0c03f1/src/Illuminate/Database/Eloquent/Model.php#L321-L371

This library currently checks for the ability to update timestamps from the Model itself:
https://github.com/iksaku/laravel-mass-update/blob/5a8bda6739b83903070f5ee9643fc88de370b886/src/MassUpdatable.php#L214-L216

It should be possible to open a PR to Laravel directly to allow this kind of configuration-based behavior and have the existing functions pickup such configuration.

I would strongly suggest looking into the possibility of integrating such feature at framework level and this library would automatically pickup your desired behavior without causing edge-cases.

**Reply by stevenmaguire** [reply]
I get it. Following the stock Laravel convention is definitely a safe way to go. The stock Laravel behavior does support this via the [HasTimestamps](https://github.com/laravel/framework/blob/9.x/src/Illuminate/Database/Eloquent/Concerns/HasTimestamps.php) trait. 

<img width="478" alt="CleanShot 2022-08-08 at 20 04 36@2x" src="https://user-images.githubusercontent.com/1851973/183540550-f3fb879a-8edf-40b8-a866-85eb4a8bbe4e.png">

<img width="464" alt="CleanShot 2022-08-08 at 20 04 56@2x" src="https://user-images.githubusercontent.com/1851973/183540556-193bc343-241a-4d3a-b31f-c40717333a25.png">

The expectation is that the configuration is operating against an instance. If you were operating against a single instance of a model, the following would disabled timestamp touching (directly, not relations):

```php
$model->timestamps = false;
$model->save(); // No updated_at modification
$model->timestamps = true;
$model->save(); // With updated_at modification
```

The trouble is arrises with this comment that decorates the method you've highlighted:

https://github.com/iksaku/laravel-mass-update/blob/5a8bda6739b83903070f5ee9643fc88de370b886/src/MassUpdatable.php#L213

In other words, "if the [AUTOMATICALLY INSTANTIATED] model (`$this`) has `$this->timestamps == true` (which it most certainly does by default, unless someone goes out of their way to override that for a model, for all of their default use cases)" then the "update it for all touched records" portion of the comment will always touch the timestamps.

This is a troubling use case, for sure and I haven't seen any other library attempt it as part of the library's API. The most popular one (at the moment) is mavinoo/laravel-batch and it steers clear of "model-attached" ergonomics (it does not furnish a static method, scope or otherwise) but still leverages a freshly instantiated model for direction on how to map data.

It stands to reason that the following would - hopefully - work for this library:

```php
$model = new Model();
$model->timestamps = false;
$model->massUpdate(
    values: [
        ['id' => 1, 'name' => 'Jorge González'],
        ['id' => 2, 'name' => 'Gladys Martínez'],
    ]
);
$model->timestamps = true;
```

I just don't see test cases to support that and, ergonomically speaking, it is a bit cumbersome. 

Perhaps a third parameter that accepts some configuration could tidy that up?

```php
Model::massUpdate(
    values: [
        ['id' => 1, 'name' => 'Jorge González'],
        ['id' => 2, 'name' => 'Gladys Martínez'],
    ],
    uniqueBy: 'username',
    options: ['timestamps' => false]
);
```

Or a fluent setter that resets after each "massUpdate" - to avoid the "turn it off" then "turn it on" convention disclosed above?

```php
Model::withoutMassUpdateTimestamp()->massUpdate(
    values: [
        ['id' => 1, 'name' => 'Jorge González'],
        ['id' => 2, 'name' => 'Gladys Martínez'],
    ],
    uniqueBy: 'username'
);
```

**Reply by cwilby** [reply]
@stevenmaguire @iksaku just an FYI - I'm currently testing a couple of approaches to perform mass updates quietly.

Here are the candidates:
```php
Model::withoutMassUpdateTimestamps()->massUpdate(...);

Model::massUpdateQuietly(...);
```

```php
trait MassUpdatable
{
    public static function withoutMassUpdateTimestamps(): self
    {
        return tap(new static(), function (self $model) {
            $model->timestamps = false;
        });
    }    
    
    public function scopeMassUpdateQuietly(Builder $query, array | Arrayable $values, array | string | null $uniqueBy = null): int
    {
        $this->timestamps = false;

        return tap($this->massUpdate($query, $values, $uniqueBy), function () {
            $this->timestamps = true;
        });
    }
}
```

**Reply by cwilby** [reply]
PR with tests: https://github.com/iksaku/laravel-mass-update/pull/6/files

@stevenmaguire this is mostly a porting of your legwork into code - thank you!

**Comment by stevenmaguire** [comment]
@cwilby awesome! that looks promising - thanks for investing the elbow grease. I'm glad to hear my chicken scratch proved useful, perhaps inspiring. It looks like you closed the PR?

**Reply by cwilby** [reply]
@stevenmaguire you're welcome!  I closed the PR after learning about `withoutTimestamps`, which was added about a month after this discussion began.

https://github.com/laravel/framework/pull/44138

```
Model::withoutTimestamps(fn () => Model::massUpdate(...));
```

Generally wanted to avoid adding a feature that is supported in Laravel.

**Reply by stevenmaguire** [reply]
I still think there is room, and use case for smoothing out the ergonomic friction with this library. Under the hood it can, and should, use Laravel's stock features. IMO the `Model::withoutTimestamps(fn () => Model::massUpdate(...));` approach is still clunky.

I think the `Model::massUpdateQuietly(...);` add here is still valuable, even if the underlying approach is different than what you previously proposed.

Your branch was right there! Perhaps a second pass to introduce that method as first-party support - with your tests! - will be acceptable?

**Reply by cwilby** [reply]
Reopened ¯\_(ツ)_/¯

**Reply by stevenmaguire** [reply]
I just added a comment!

**Reply by iksaku** [reply]
PR merged and tagged the `1.0.5` release 🥳

**Reply by stevenmaguire** [reply]
![huzzah!](https://media.giphy.com/media/zdXPfSo1UdRu3oqu1C/giphy.gif)

Thank you @iksaku and @cwilby - great work!
